### ISSUE TITLE
Support for OIDC distributed claims

### ISSUE THREAD

**Issue by jessesuen** [argoproj/argo-cd#7127]
# Summary

OIDC providers, namely Azure AD, make use of OIDC [distributed claims](https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims). This feature in the OIDC spec, separates out the claims information into a separate URL endpoint. This endpoint is expected to be visited by the OIDC client app, should they need to inspect that information (i.e. for auth-z).

Here is an example of a JWT using distributed claims:

```json
  {
   "name": "Jane Doe",
   "given_name": "Jane",
   "family_name": "Doe",
   "email": "janedoe@example.com",
   "birthdate": "0000-03-22",
   "eye_color": "blue",
   "_claim_names": {
     "payment_info": "src1",
     "shipping_address": "src1",
     "credit_score": "src2"
    },
   "_claim_sources": {
     "src1": {"endpoint":
                "https://bank.example.com/claim_source"},
     "src2": {"endpoint":
                "https://creditagency.example.com/claims_here",
              "access_token": "ksj3n283dke"}
   }
  }
```

In the case of Azure AD, when a user reaches 200 groups, the Azure claim will switch to returning distributed claims in the JWT. See [docs](https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims):

> In larger organizations the number of groups a user is a member of may exceed the limit that Azure Active Directory will add to a token. 150 groups for a SAML token, and 200 for a JWT. This can lead to unpredictable results. If your users have large numbers of group memberships, we recommend using the option to restrict the groups emitted in claims to the relevant groups for the application.

From the perspective of Argo CD, it is as if the user suddenly is part of *no* groups, and so will have no permissions in the.

# Motivation

Azure AD is a popular IdP and we should have better support for it. Without this feature, users have to configure the OIDC Client App to specifically choose which claims are returned in claims. This process is not very scalable, or obvious.

# Proposal

As soon as the OIDC flow is completed, and the Argo CD should immediately make a follow up request to the claims sources which contains the group information.

Here is Vault's auth plugin implementation for handling distributed claims which we should take inspiration from:

https://github.com/hashicorp/vault-plugin-auth-jwt/blob/f8f6db66e0f54d2d003f2c796fe920814d86edb0/provider_azure.go#L74

There are two workarounds which I can think of:

1. Azure AD can be configured such that it only returns a limited amount of cherry-picked groups to return in the claims. However, this adds a burden that any time you want a new group to be considered for authz, you would need to go to Azure AD and add the group UUID to the list of groups to return to the Argo CD client app JWT claims.
2. For users who exceeds 200 groups, a special entry in argocd-rbac-cm could be added such that something *other* than their groups. (e.g. their email address) gets mapped to a special role. However this approach may only work or scale for users who are special (e.g. admins).

**Comment by jessesuen** [comment]
Here is what I found. Microsoft Azure AD is completely broken & non-compliant with regards to OIDC.

Here is an example of a JWT "distributed claims" returned by Azure using their "OIDC" endpoints, e.g. 
`https://login.microsoftonline.com/11111111-1111-1111-1111-111111111111/v2.0`

```json
{
   "_claim_names":{
      "groups":"src1"
   },
   "_claim_sources":{
      "src1":{
         "endpoint":"https://graph.windows.net/11111111-1111-1111-1111-111111111111/users/22222222-2222-2222-2222-222222222222/getMemberObjects"
      }
   },
   "aud":"c6eb3724-4028-11ec-9356-0242ac130003",
   "email":"user@example.com",
   "exp":1636333467,
   "iat":1636329567,
   "idp":"https://sts.windows.net/d58f9b6c-4028-11ec-9356-0242ac130003/",
   "iss":"https://login.microsoftonline.com/11111111-1111-1111-1111-111111111111/v2.0",
   "name":"Hong Wang",
   "nbf":1636329567,
   "oid":"22222222-2222-2222-2222-222222222222",
   "preferred_username":"user@example.com",
   "rh":"...",
   "sub":"...",
   "tid":"11111111-1111-1111-1111-111111111111",
   "uti":"...",
   "ver":"2.0",
   "wids":[
      "dbcae09a-4028-11ec-9356-0242ac130003",
      "dfc72adc-4028-11ec-9356-0242ac130003"
   ]
}
```

Here are all the ways they are non-compliant that I have found:

**1. The client must **_not_** request the `groups` scope even if you configured Azure to return it**

Azure does not like the client application to request the "groups" scope, even though that is the very thing that is configured in the IDP to return. Doing so will cause Azure to error with:

```
The application 'Argo CD' asked for scope 'groups' that doesn't exist on the resource '00000003-0000-0000-c000-000000000000'. Contact the app vendor.
```

To work around this, I configured Argo CD to explicitly not request groups:

```yaml
data:
  requestedScopes: ["openid", "profile", "email"]
```

Azure will return `groups` anyways if you configure it in the token.

**2. The endpoint specified in the `_claim_sources` is unusable as is** 

If you attempt to visit the endpoint in the `_claim_sources` as is, you will receive the error:

```json
{
    "odata.error": {
        "code": "Request_DataContractVersionMissing",
        "message": {
            "lang": "en",
            "value": "The specified api-version is invalid. The value must exactly match a supported version."
        }
    }
}
```

This is because Azure expects you to append an `api-version` query param (e.g. `?api-version=1.6`) to the endpoint. But rather than doing that, you are better off switching their newer graph v1.0 endpoint.  So instead of:

`https://graph.windows.net/11111111-1111-1111-1111-111111111111/users/22222222-2222-2222-2222-222222222222/getMemberObjects`

You would instead use:
`https://graph.microsoft.com/v1.0/11111111-1111-1111-1111-111111111111/users/22222222-2222-2222-2222-222222222222/getMemberObjects`

However, even their graph v1.0 endpoint is broken as you will see below:

**3. Azure's claims endpoint requires a POST with body, instead of GET**

OIDC distributed claims endpoints are normally retrieved with an HTTP GET request. Instead Azure requires the `getMemberObjects` call to be a POST with the following body:

```json
{
    "securityEnabledOnly": false
}
```

See: https://docs.microsoft.com/en-us/graph/api/user-getmemberobjects?view=graph-rest-1.0&tabs=http

**4. The endpoint returns JSON and not a JWT**

OIDC spec states: "The endpoint URL MUST return the Claim as a JWT."  See: https://openid.net/specs/openid-connect-core-1_0.html#AggregatedDistributedClaims

Azure does not do this. Instead, it returns something like:

```json
{
    "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#Collection(Edm.String)",
    "value": [
        "9f6792a4-52da-4508-bca8-bcf3fb27c216",
        "0de0a4aa-d5df-4796-8808-4beb222956de",
...
        "f6e0505a-130a-4ab7-89cd-79eab0f6565c"
    ]
}
```
which are the UUIDs of the groups.

In summary, there are really two changes that could be made to Argo CD:

1. Support for OIDC distributed claims for *compliant* OIDC identity providers (e.g. not Azure)
2. Special case Azure AD handling for their non-compliant OIDC-like graph v1.0 endpoints

I don't really like the idea of implementing the special case handling for 2, especially because that there may be OIDC compliant implementation in Azure in the future.

**Comment by jessesuen** [comment]
I have a working PoC of the OIDC authentication flow working with special case handling of Azure:
https://github.com/jessesuen/argo-cd/tree/oidc-distributed-claims

Still missing:
1. a storage strategy of what to do with the distributed claims after they are fetched (e.g. cookie? redis? a new, merged JWT?)
2. multi-JWT auth processing in argo-cd API server
3. a cleaner way to deal with Azure AD / Microsoft Graph. I don't want to lump this in with Argo CD's native OIDC configuration/implementation because Azure is not compliant. So if we do this, we would need to be a separate `azure` auth provider and configuration.

**Comment by hussainbani** [comment]
We have faced similar issue, if the AD groups increases more than 200 groups, its send you distributed claims.

Is there a way to tell argo to use distributed claims instead?

**Comment by gravufo** [comment]
@jessesuen the v2 endpoint is supposed to be fully compliant: https://github.com/kubernetes/kubernetes/issues/62920#issuecomment-443060890
Is there a reason you're implementing it for the v1 endpoint? I think most people can easily use v2 and it can be considered a limitation that v1 does not support distributed claims in ArgoCD.

**Comment by gravufo** [comment]
I've opened a PR to address this. @jessesuen would it be possible to review the PR and let me know what you think? I know you wanted to prevent modifying the oidc provider directly, but at this point I think Azure is a major OIDC provider, and first class support is expected, especially since Azure is already supported and extensively documented in ArgoCD.

I agree that Microsoft is not great by doing things non-oidc compliant, but since the config option to enable this is azure-specific (under the azure key in the auth config) and the code branching is based on that, I feel it's safe enough to proceed and doesn't impact oidc-compliant providers.

Since this issue has been open for almost 5 years now, I hope that we can get a working fix in ArgoCD soon :)

**Comment by hussainbani** [comment]
> I've opened a PR to address this. [@jessesuen](https://github.com/jessesuen) would it be possible to review the PR and let me know what you think? I know you wanted to prevent modifying the oidc provider directly, but at this point I think Azure is a major OIDC provider, and first class support is expected, especially since Azure is already supported and extensively documented in ArgoCD.
> 
> I agree that Microsoft is not great by doing things non-oidc compliant, but since the config option to enable this is azure-specific (under the azure key in the auth config) and the code branching is based on that, I feel it's safe enough to proceed and doesn't impact oidc-compliant providers.
> 
> Since this issue has been open for almost 5 years now, I hope that we can get a working fix in ArgoCD soon :)

Do we know when this can be merged, we are waiting for almost 2 years for this feature.
